bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java

```diff
@@ -594,8 +608,57 @@ public BookieNode selectFromNetworkLocation(
         }
     }
 
-    protected String getRemoteRack(BookieNode node) {
-        return "~" + node.getNetworkLocation();
+    protected BookieNode selectFromNetworkLocation(String networkLoc,
+                                                   Set<String> excludeRacks,
+                                                   Set<Node> excludeBookies,
+                                                   Predicate<BookieNode> predicate,
+                                                   Ensemble<BookieNode> ensemble)
+            throws BKNotEnoughBookiesException {
+        // first attempt to select one from local rack
+        try {
+            return selectRandomFromRack(networkLoc, excludeBookies, predicate, ensemble);
+        } catch (BKNotEnoughBookiesException e) {
+            if (isWeighted) {
+                // if weight based selection is enabled, randomly select one from the whole cluster
+                // based on weights and ignore the provided predicate.
```

Comment:
"ignore the provided predicate" -> maybe the comment is wrong ?  it seems to me that the predicate is applied in selectRandom